### PR TITLE
Debugging lambdas in mfem::forall

### PR DESCRIPTION
**Pull request by camierjs**
Should address #2958.
Checks fail because of "Number of files changed in one commit (94) exceeds maximum allowable (50)."
___

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3454](https://github.com/mfem/mfem/pull/3454) | @camierjs | @tzanio | @artv3 + @YohannDudouit + @tomstitt + @v-dobrev | 2/11/23 | 4/14/23 | 4/18/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by artv3 — APPROVED** [review]
yes!

**Comment by samuelpmishLLNL** [comment]
This seems like a positive change to me. I didn't see where the `mfem::forall**` overloads were actually defined, but I might have missed it given the number of files touched in this PR.

**Review by samuelpmishLLNL — APPROVED** [review]

**Comment by camierjs** [comment]
> This seems like a positive change to me. I didn't see where the `mfem::forall**` overloads were actually defined, but I might have missed it given the number of files touched in this PR.

Thank you for the review @samuelpmishLLNL , the `forall` [changes](https://github.com/mfem/mfem/blob/1cd062b77f87b26ad0162fedcebd4009110561ae/general/forall.hpp#L620-L651) are in the `general/forall.hpp` file.

File: general/forall.hpp [hunk 1]
```diff
@@ -61,38 +61,33 @@ const int MAX_Q1D = 14;
 // The MFEM_FORALL wrapper
 #define MFEM_FORALL(i,N,...)                             \
    ForallWrap<1>(true,N,                                 \
-                 [=] MFEM_DEVICE (int i) {__VA_ARGS__},  \
-                 [&] MFEM_LAMBDA (int i) {__VA_ARGS__})
+                 [=] MFEM_DEVICE (int i) {__VA_ARGS__})
```
Comment (tomstitt) [inline general/forall.hpp:64]: Is there a reason this isn't `MFEM_HOST_DEVICE` ?

**Review by tomstitt — APPROVED** [review]
thank you @camierjs !

File: general/forall.hpp (hunk 1 above)
Comment (camierjs) [reply general/forall.hpp:64]: You're right, it should be `MFEM_HOST_DEVICE`: I'll update it, thank you @tomstitt

File: fem/dgmassinv.hpp [hunk 2]
```diff
@@ -102,7 +102,7 @@ class DGMassInverse : public Solver
 
    /// @brief Solve the system M b = u. <b>Not part of the public interface.</b>
    /// @note This member function must be public because it contains an
-   /// MFEM_FORALL kernel (nvcc limitation)
+   /// mfem::forall kernel (nvcc limitation)
```
[the remark this answers is not in the document]
Comment (camierjs) [reply fem/dgmassinv.hpp]: I've updated the comment note.

**Comment by artv3** [comment]
Does this change effect performance? Or is it pretty much the same?

**Comment by camierjs** [comment]
> Does this change effect performance? Or is it pretty much the same?

They are exactly the same on GPU (no changes there), and we did some benchmarks on the CPU without noticing any performance loss.

**Review by YohannDudouit — APPROVED** [review]
Tested with an intentional SegFault, that works very well, we get the exact line for the issue in a debugger. I think this is a big improvment for users.

**Comment by tzanio** [comment]
@v-dobrev, can you please take a look so we can move forward with this (there are 4 other approvals already)

File: CHANGELOG [hunk 3]
```diff
@@ -65,6 +65,9 @@ Version 4.5.2, released on March 23, 2023
 - The following integrations have updated minimum version requirements:
   * RAJA >= 2022.10.3
 
+- Adds mfem::forall functions that can be used inplace of MFEM_FORALL macros,
+  which have been kept for backward compatibility.
+
```
Comment (v-dobrev) [inline CHANGELOG:70]: Move to the section for v4.5.3

Also, it is good to clarify that the new function allows for better debugging of lambda bodies compared to the `MFEM_FORALL` macros.

File: fem/bilininteg_hcurl.cpp [hunk 4]
```diff
@@ -1950,12 +1950,7 @@ static void SmemPACurlCurlApply3D(const int D1D,
       } // qz
    }; // end of element loop
 
-   auto host_kernel = [&] MFEM_LAMBDA (int)
-   {
-      MFEM_ABORT_KERNEL("This kernel should only be used on GPU.");
-   };
-
-   ForallWrap<3>(true, NE, device_kernel, host_kernel, Q1D, Q1D, Q1D);
+   ForallWrap<3>(true, NE, device_kernel, Q1D, Q1D, Q1D);
```
Comment (v-dobrev) [inline fem/bilininteg_hcurl.cpp:1953]: We should be careful here: I suspect this change will revert some fixes in #2788 for issue #2723. We need to find a new solution for the warnings.

File: fem/bilininteg_hcurl.cpp [hunk 5]
```diff
@@ -4208,12 +4203,7 @@ static void SmemPAHcurlL2Apply3D(const int D1D,
...
```
Comment (v-dobrev) [inline fem/bilininteg_hcurl.cpp:4206]: Similar to the above comment.

File: fem/bilininteg_hcurl.cpp [hunk 6]
```diff
@@ -5670,12 +5660,7 @@ static void SmemPAHcurlL2Apply3DTranspose(const int D1D,
...
```
Comment (v-dobrev) [inline fem/bilininteg_hcurl.cpp:5663]: Similar to the above comment.

File: fem/bilininteg_hcurl.cpp (hunk 4 above)
Comment (camierjs) [reply fem/bilininteg_hcurl.cpp]: I reverted and allow back the `device` and `host` lambdas in [aa4b04d](https://github.com/mfem/mfem/pull/3454/commits/aa4b04d0a8711e5506191a212645ce26d8f9328a)

Comment (v-dobrev) [reply fem/bilininteg_hcurl.cpp]: This should work. Thanks!

**Review by v-dobrev — APPROVED** [review]

**Pull request merged**